core/win32/drwinapi/ntdll_redir.c

```diff
@@ -838,15 +838,27 @@ redirect_RtlPcToFileHeader(__in PVOID PcValue, __out PVOID *BaseOfImage)
  * i#875: FLS isolation
  */
 
+#ifndef FLS_MAX_COUNT
+#    define FLS_MAX_COUNT 128
+#endif
+
 void
 ntdll_redir_fls_init(PEB *app_peb, PEB *private_peb)
 {
     /* FLS is supported in WinXP-64 or later */
     ASSERT(get_os_version() >= WINDOWS_VERSION_2003);
+
+    /* FIXME i#3633: Implement FLS isolation in Win10 1903+ where FLS data is no longer
```

Comment:
And we want to change this s/FIXME/XXX/ since nice-to-have instead of must-have: and should be reworded to sthg like "XXX i#3633: On Win10 1903+, better to keep our data in the ntdll vars that Windows uses instead of the PEB", right?

Reply:
This is something which I had in mind to isolate it from PEB from all windows versions:

https://github.com/deroko/dynamorio/commit/b62633fa733259f6a27a0172063c507335131c37

Reply:
Seems a bit bigger fix then what I have expected. 
PEB is isolated, that is perfectly fine, but in private_peb there is FlsBitmap used to locate empty slots in FLS, which is in teb->FlsData. Application on the other hand uses app_peb->FlsBitmap, but also stores data in teb->FlsData, so there is a collision between these 2. Probably problem never happened as nobody noticed it, or FLS has 0 real usage.

How I've fixed it, is that I use same trick as for sysenter handling to allocate TLS slot, where private FlsData will be held. There should be no more collision between app/DynamoRIO FLS handling.

This is done by tls_alloc() to reserve TLS slot for private FlsData. I'm wondering if it is possible to extend dcontext struct to have one extra field, thus there is no need to occupy one extra TLS slot on Windows?

This branch has all of the above mentioned:
https://github.com/deroko/dynamorio/commits/fix-3633-ver2

Any advice from code style/organization how to continue? Shall I use tls provided by DynamoRIO via `tls_alloc` or to extend drcontext_t->client_data to have one extra field just for this purpose, and not to waste one extra TLS in TEB?

Reply:
My bad, found code which isolates TEB FlsData in win32/loader.c from `swap_peb_pointer:713`:
`
if (dcontext->priv_fls_data != cur_fls) { /* handle two calls in a row */
                    dcontext->app_fls_data = cur_fls;
                    set_teb_field(dcontext, FLS_DATA_TIB_OFFSET, dcontext->priv_fls_data);
                }
`

Good, fix is now easier.

If it is possible to close this pull request, I'll open a new one with proper fix.